presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -289,6 +305,73 @@ private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootB
     }
 
     private List<SqlStageExecution> createStages(
+            ExchangeLocationsConsumer locationsConsumer,
+            LocationFactory locationFactory,
+            StreamingPlanSection section,
+            Optional<int[]> bucketToPartition,
+            OutputBuffers outputBuffers,
+            NodeScheduler nodeScheduler,
+            RemoteTaskFactory remoteTaskFactory,
+            SplitSourceFactory splitSourceFactory,
+            Session session,
+            int splitBatchSize,
+            NodePartitioningManager nodePartitioningManager,
+            ExecutorService queryExecutor,
+            ScheduledExecutorService schedulerExecutor,
+            FailureDetector failureDetector,
+            NodeTaskMap nodeTaskMap,
+            ImmutableMap.Builder<StageId, StageScheduler> stageSchedulers,
+            ImmutableMap.Builder<StageId, StageLinkage> stageLinkages)
+    {
+        ImmutableList.Builder<SqlStageExecution> stages = ImmutableList.builder();
+
+        // Only fetch a distribution once per section to assure all stages see the same machine assignments
```

Comment:
nit: s/assure/ensure

Reply:
also, this is to keep to how many resources we want to give the query, right? it's not a correctness issue.

Reply:
@rschlussel : It 's to make sure the same bucket assign to the same node in one query (not "section"). For example, thinking about 

```
        JOIN
        /    \
 TS(1024) TS(1024)
```

The Hive partitioning only defines how to bucket the data, but doesn't define the bucket to node map. The bucket to node map was decided in `NodePartitioningMap#getNodePartitioningMap` by randomly shuffle:  https://github.com/prestodb/presto/blob/b88174803cf4a1d3b36918ac91db2465b350ac90/presto-main/src/main/java/com/facebook/presto/sql/planner/NodePartitioningManager.java#L146-L148 

We must make sure the bucket to node map is same in the same section (and thus we have this `partitioningCache` reused during scheduling.